modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+
+        Quat<_Tp> operator/(const Quat<_Tp>&) const;
+
+        Quat<_Tp>& operator/=(const _Tp&);
+
+        Quat<_Tp>& operator/=(const Quat<_Tp>&);
+
+        _Tp& operator[](std::size_t n);
+
+        const _Tp& operator[](std::size_t n) const;
+
+        template <typename S, typename T>
+        friend Quat<S> cv::operator*(const T, const Quat<S>&);
+
+        template <typename S, typename T>
+        friend Quat<S> cv::operator*(const Quat<S>&, const T);
+
+        template <typename S>
+        friend std::ostream& cv::operator<<(std::ostream&, const Quat<S>&);
+
+        _Tp w, x, y, z;
+
+    };
+    template <typename T>
+    Quat<T> rvec2Quat(Vec<T, 3> &rvec);
+
+    template <typename T>
+    Quat<T> inv(const Quat<T> &q, bool assumeUnit=false);
+
+    template <typename T>
+    Quat<T> sinh(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> cosh(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> tanh(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> sin(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> cos(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> tan(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> asinh(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> acosh(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> atanh(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> asin(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> acos(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> atan(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> power(const Quat<T> &q, const Quat<T> &p, bool assumeUnit=false);
+
+    template <typename T>
+    Quat<T> exp(const Quat<T> &q);
+
+    template <typename T>
+    Quat<T> log(const Quat<T> &q, bool assumeUnit=false);
+
+    template <typename T, typename _T>
+    Quat<T> power(const Quat<T>& q, _T x, bool assumeUnit=false);
+
+    template <typename T>
+    Quat<T> crossProduct(const Quat<T> &p, const Quat<T> &q);
+
+    template <typename S>
+    Quat<S> sqrt(const Quat<S> &q, bool assumeUnit=false);
+
+    template <typename S, typename T>
+    Quat<S> operator*(const T, const Quat<S>&);
+
+    template <typename S, typename T>
+    Quat<S> operator*(const Quat<S>&, const T);
+
+    template <typename S>
+    std::ostream& operator<<(std::ostream&, const Quat<S>&);
+
+    using Quatd = Quat<double>;
+    using Quatf = Quat<float>;
+    using std::sin;
+    using std::cos;
+    using std::tan;
+    using std::exp;
+    using std::log;
+    using std::sinh;
+    using std::cosh;
+    using std::tanh;
+    using std::sqrt;
+    using std::asinh;
+    using std::acosh;
+    using std::atanh;
+    using std::asin;
+    using std::acos;
+    using std::atan;
+//! @} core
+}
+
+//! @cond IGNORE
+///////////////////////////////////////////////////////////////////////////////////////
+//Implementation
+namespace cv{
+    template <typename T>
```

Comment:
Please move implementation into `quaternion.inl.hpp` file and include it here.
(to avoid issues with documentation or bindings in the future).

Reply:
OK